applications/coreutils/tac.cpp

```diff
@@ -0,0 +1,123 @@
+#include <libsystem/cmdline/CMDLine.h>
+#include <libsystem/io/Path.h>
+#include <libsystem/io/Stream.h>
+#include <libutils/String.h>
+#include <libutils/Vector.h>
+
+static bool before = false;
+static char *separator = nullptr;
+
+static const char *usages[] = {
+    "NAME...",
+    "[OPTION]... NAME...",
+    nullptr,
+};
+
+static CommandLineOption options[] = {
+    COMMANDLINE_OPT_HELP,
+    COMMANDLINE_OPT_STRING("separator", 's', separator,
+                           "choose the separator to be used instead of \\n",
+                           COMMANDLINE_NO_CALLBACK),
+    COMMANDLINE_OPT_BOOL("before", 'b', before,
+                         "attach the separator before instead of after the string",
+                         COMMANDLINE_NO_CALLBACK),
+    COMMANDLINE_OPT_END};
+
+static CommandLine cmdline = CMDLINE(
+    usages,
+    options,
+    "Concatenate and print lines of file(s) in reverse.",
+    "If no filename provided read from input stream");
+
+Result tac(const char *path)
+{
+
+    __cleanup(stream_cleanup) Stream *stream = stream_open(path, OPEN_READ);
+
+    if (handle_has_error(stream))
+    {
+        return handle_get_error(stream);
+    }
```

Comment:
Why not open the file outside of the tac() function and pass the stream as a parameter. This will give you stdin support for free :)